File: DataStructure/AVLTree.py
class Node:
    def __init__(self, value):
        self.value = None
        self.left = None
        self.right = None
        self.height = None


def Height(node):
    if node is None:
        return -1
    else:
        return node.height


def Find(x, root):
    if root is None or root.value == x:
        return root
    elif root.value < x:
        if root.left is None:
            return None
        else:
            return Find(x, root.left)
    elif root.vlaue > x:
        if root.right is None:
            return None
        else:
            return Find(x, root.right)


def FindMin(root):
    if root is None:
        return None
    elif root.left is None:
        return root
    else:
        return FindMin(root.left)


def FindMax(root):
    if root is None:
        return None
    else:
        while root.right is not None:
            root = root.right
        return root


def SingleLeft(node):
    root = node.left
    node.left = root.right
    root.right = node.right
    root.height = max(Height(root.left), Height(root.right)) + 1
    node.height = max(Height(root.left), Height(root.right)) + 1
    return root


def SingleRight(node):
    root = node.right

    return root


def DoubleLeft(node):
    node.left = SingleRight(node.left)

    return SingleLeft(node)


def DoubleRight(node):
    node.right = SingleLeft(node.right)

    return SingleRight(node)


def Insert(x, root):
    if root is None:
        root = Node()
        root.value = x
        return root
    else:
        if root.value > x:
            root.left = Insert(x, root.left)
            if Height(root.left) - Height(root.right) == 2:
                if x < root.left.value:
                    root = SingleLeft(root)
                else:
                    root.left = DoubleLeft(root)
        elif root.value < x:
            root.right = Insert(x, root.right)
            if Height(root.right) - Height(root.left) == 2:
                if x > root.right.value:
                    root = SingleRight(root)
                else:
                    root.right = DoubleRight(root)
    root.height = max(Height(root.left), Height(root.right)) + 1
    return root


def Delete(x, root):
    if root is None:
        return None
    else:
        if root.value == x:
            if root.right is not None and root.left is not None:
                root.value = FindMin(root.right)
            else:
                if root.right is None:
                    root = root.left
                elif root.left is None:
                    root = root.rigth
    if root is not None:
        # TODO
        pass

    return root


if __name__ == '__main__':
    pass

File: DataStructure/AVLTree.java
public class AVLTree{
    private Node root = null;

    public void AVLTree(){}
    
    public void AVLTree(Node root){
        this.root = root;
    }

    public Node getRoot(){
        return this.root;
    }

    public void insert(int x){
        if(root == null){
            Node v = new Node(x);
            root = v;
        }else{
            insertNode(x, root);
        }
    }

    public Node FindMin(Node root){
        if(root == null){
            return null;
        }else if(root.left == null){
            return root;
        }else{
            return FindMin(root.left);
        }
    }

    public Node FindMax(Node root){
        if(root == null){
            return null;
        }else if(root.right == null){
            return root;
        }else{
            return FindMax(root.right);
        }
    }

    public Node Find(int x){
        Node temp = root;
        while(temp != null){
            if(temp.val == x){
                break;
            }else{
                if(temp.val > x){
                    temp = temp.left;
                }else{
                    temp = temp.right;
                }
            }
        }
        return temp;
    }

    private Node insertNode(int x,Node root){
        if(root == null){
            return new Node(x);
        }else{
            if(root.val > x){
                root.left = insertNode(x, root.left);
                if(getHeight(root.left) - getHeight(root.right) == 2){
                    // TODO
                    if(root.left.val > x){
                        root = singleRotateLeft(root);
                    }else{
                        root = doubleRotateLeft(root);
                    }
                }

            }else{
                root.right = insertNode(x, root.right);
                if(getHeight(root.right) - getHeight(root.left) == 2){
                    // TODO
                    if(root.right.val < x){
                        root = singleRotateRight(root);
                    }else{
                        root = doubleRotateRight(root);
                    }
                }
            }
            root.height = Math.max(getHeight(root.left), getHeight(root.right)) + 1;
            return root;   
        }
    }

    public Node delete(int x){
        if(root != null){
            if(root.val == x){
                if(root.left == null && root.right == null){
                    root = null;
                }
                else if(root.left != null && root.right == null){

                }
            }
        }
        return root;
    }

    private Node singleRotateLeft(Node n){
        Node ret = n.left;
        n.left = ret.right;
        ret.right = n;
        ret.height = Math.max(getHeight(ret.left),getHeight(ret.right)) + 1;
        n.height = Math.max(getHeight(n.left),getHeight(n.night)) + 1;
        return ret;
    }

    private Node singleRotateRight(Node n){
        Node ret = n.right;
        n.right = ret.left;
        ret.left = n;
        ret.height = Math.max(getHeight(ret.left),getHeight(ret.right)) + 1;
        n.height = Math.max(getHeight(n.left),getHeight(n.night)) + 1;
        return ret;
    }

    private Node doubleRotateLeft(Node n){
        n.left = singleRotateRight(n.left);
        return singleRotateLeft(n);
    }

    private Node doubleRotateRight(Node n){
        n.right = singleRotateLeft(n.right);
        return singleRotateRight(n);
    }

    public int getHeight(Node n){
        if(n == null){
            return 0;
        }else{
            return 1 + Math.max(getHeight(n.left),getHeight(n.right));
        }
    }
}


class Node{
    int val;
    int height = 0;
    Node left = null;
    Node right = null;
    public void Node(int val){
        this.val = val;
    }

    public int getHeight(){
        return this.height;
    }
}